src/SHA/Keccak.mo

```diff
@@ -0,0 +1,242 @@
+        private var pt: Nat = 0;
+
+        // blockSize in bytes
+        public func blockSize() : Nat { rsize; };
+
+        // Function to initialize
+        public func reset() : () {
+            for (i in Iter.range(0, state.size()-1)) {
+                state[i] := 0;
+            };
+            pt := 0;
+        };
+
+        // Return size of hash in bits
+        public func size() : Nat { hashSize; };
+
+        // Turn array of Nat8s into Nat64, LSB first
+        private func pack64(data : [Nat8]) : Nat64 {
+            let dsize = data.size();
+            assert(dsize <= 8 and dsize > 0);
+
+            var q: Nat64 = 0;
+            // Avoid reverse() here to optimize? Maybe unroll?
+            for (i in Iter.range(0, dsize-1:Nat)) {
+                q |= Nat64.fromNat(Nat8.toNat(data[dsize-1:Nat-i]));
+                if (i < (dsize-1:Nat)) {
+                    q <<= 8;
+                };
+            };
+            return q;
+        };
+
+        // Unpack Nat64 into array of Nat8s, LSB first
+        private func unpack64(q : Nat64) : [Nat8] {
+            var t = q;
+            let qbuf = Array.init<Nat8>(8, 0);
+            for (i in Iter.range(0, qbuf.size()-1)) {
+                qbuf[i] := Nat8.fromNat(Nat64.toNat(t & 0xff));
+                t >>= 8;
+            };
+            return Array.freeze<Nat8>(qbuf);
+        };
+
+        private func dump() {
+            Debug.print("State");
+            for (i in Iter.range(0, state.size()-1)) {
+                let b = Array.reverse<Nat8>(unpack64(state[i]));
+                Debug.print(Hex.encode(b));
+            };
+        };
+
+        // Repeated from SHA2. Candidate for DRY principle refactor?
+        public func sum(bs : [Nat8]) : [Nat8] {
+            let cs = checkSum();
+            let size = bs.size();
+            Array.tabulate<Nat8>(
+                size + cs.size(),
+                func (x : Nat) {
+                    if (x < size) return bs[x];
+                    cs[x - size];
+                }
+            );
+        };
+
+        // Function to finalize hashing
+        public func checkSum() : [Nat8] {
+            state[pt/8] ^= Nat64.fromNat(Nat8.toNat(delimitedSuffix)) << Nat64.fromNat((pt%8)*8);
+            state[(rsize-1:Nat)/8] ^= 0x80 : Nat64 << Nat64.fromNat(((rsize-1):Nat%8)*8);
+
+            //dump();
+            keccak_f(state);
+
+            let md = Array.init<Nat8>(hashSize/8, 0);
+
+            var i = 0;
+            label done for (q in state.vals()) {
+                for (b in unpack64(q).vals()) {
+                    if (i >= md.size()) {
+                        break done;
+                    };
+                    md[i] := b;
+                    i += 1;
+                };
+            };
+            return Array.freeze<Nat8>(md);
+        };
+
+        // Function to update internal state with content
+        // TODO: should data be a blob? Sticking with Array from Hash type
```

Comment:
Once you can get individual bytes out of a blob, the whole package will migrate to it.
Currently blob do not allow you to do much.

Reply:
Good to know. I'll drop the comment.